### PR TITLE
Add Depletion ID Generation Capability into RGMB

### LINKED ISSUE #25396: Add Depletion ID Generation Capability into RGMB
## Reason

This issue is to add the depletion ID generation capability into  the `RGMB` (reactor geometry mesh build) of reactor module.
 
## Design

The depletion ID generation capability current available in `DepletionIDGenerator` will be integrated into the `RGMB`.
A user can setup the depletion ID assignment option as input parameters of the `AssemblyMeshGenerator` or the `CoreMeshGenerator` of RGMB. `DepletionIDGenerator` does not need to be included in the mesh generation process. 
```
  [core]
    type = CoreMeshGenerator
    ...
    pattern = ' 1 1;
               1 0 1;
                1 1'
    generate_depletion_id = true
    depletion_id_type = 'pin' or 'pin_type'
  []
```
## Impact

A user can assign depletion IDs during the RGMB mesh generation process.

### PR DESCRIPTION
**Pull request by yjung-anl**
closes #25396

### CODE REVIEW

**Comment by yjung-anl** [comment]
Working in progress.

**Comment by NamjaeChoi** [comment]
Will this be available very soon?

**Comment by yjung-anl** [comment]
@GiudGiud 
If you have the time, could you please review this?

**Comment by GiudGiud** [comment]
Will do. I m at ANS right now but I ll have time Friday.

File: modules/reactor/doc/content/source/meshgenerators/AssemblyMeshGenerator.md [hunk 1]
```diff
@@ -26,6 +26,15 @@ As mentioned above, the `AssemblyMeshGenerator` object will tag all elements (th
...
 
+## Depletion ID Information
+
+The `AssemblyMeshGenerator` object optionally assign depletion ID, with the extra integer name "depleiton_id", if it is in the final mesh generation level.
+The depletion ID generation option can be enabled by setting the  [!param](/Mesh/CoreMeshGenerator/generate_depletion_id) to true.
+The level of detail needed for depletion zones is specified in the input parameter [!param](/Mesh/CoreMeshGenerator/depletion_id_type).
+[!param](/Mesh/CoreMeshGenerator/depletion_id_type) can be either `pin` and `pin_type`.
+All pins in the core have separate depletion ID values by setting [!param](/Mesh/CoreMeshGenerator/depletion_id_type) to `pin`.
```
Comment (GiudGiud) [inline modules/reactor/doc/content/source/meshgenerators/AssemblyMeshGenerator.md:35]:
```suggestion
All pins in the assembly have separate depletion ID values by setting [!param](/Mesh/CoreMeshGenerator/depletion_id_type) to `pin`.
```

File: modules/reactor/doc/content/source/meshgenerators/AssemblyMeshGenerator.md [hunk 2]
```diff
@@ -26,6 +26,15 @@ As mentioned above, the `AssemblyMeshGenerator` object will tag all elements (th
 
 The `AssemblyMeshGenerator` object also automatically tags all elements in the mesh with the [!param](/Mesh/AssemblyMeshGenerator/assembly_type) using the extra_integer name "assembly_type_id" and, if extruded, elements in each axial layer are tagged the axial layers using the name "plane_id". The pins composing the assembly are also tagged via [`PatternedCartesianMeshGenerator`](PatternedCartesianMeshGenerator.md) or [`PatternedHexMeshGenerator`](PatternedHexMeshGenerator.md), using the "cell" assignment type, with the extra integer name "pin_id".
 
+## Depletion ID Information
+
+The `AssemblyMeshGenerator` object optionally assign depletion ID, with the extra integer name "depleiton_id", if it is in the final mesh generation level.
+The depletion ID generation option can be enabled by setting the  [!param](/Mesh/CoreMeshGenerator/generate_depletion_id) to true.
```
Comment (GiudGiud) [inline modules/reactor/doc/content/source/meshgenerators/AssemblyMeshGenerator.md:32]: adapt all parameters this is not the Core

File: modules/reactor/doc/content/source/meshgenerators/CoreMeshGenerator.md [hunk 3]
```diff
@@ -18,6 +18,17 @@ The `CoreMeshGenerator` object adopts much of the existing input structure of pa
 
 The `CoreMeshGenerator` object automatically tags the mesh, if three dimensional, with the axial layers using the extra integer name "plane_id". The assemblies composing the core are also tagged via [`PatternedCartesianMeshGenerator`](PatternedCartesianMeshGenerator.md) or [`PatternedHexMeshGenerator`](PatternedHexMeshGenerator.md), using the "cell" assignment type, with the extra integer name "assembly_id" and any "dummy" assembly (identified via the [!param](/Mesh/CoreMeshGenerator/dummy_assembly_name) parameter) locations excluded.
 
+## Depletion ID Information
+
+The `CoreMeshGenerator` object optionally assign depletion ID, with the extra integer name "depleiton_id".
```
Comment (GiudGiud) [inline modules/reactor/doc/content/source/meshgenerators/CoreMeshGenerator.md:23]:
```suggestion
The `CoreMeshGenerator` object can optionally assign a depletion ID, with the extra integer name "depletion_id".
```

File: modules/reactor/src/meshgenerators/ReactorGeometryMeshBuilderBase.C [hunk 4]
```diff
@@ -254,3 +268,50 @@ ReactorGeometryMeshBuilderBase::print2dMetadataToConsole(const std::string metad
...
+    id_names.push_back("plane_id");
+  id_names.push_back("region_id");
+  // prepare list of blocks
+  std::set<SubdomainID> block_ids;
+  for (Elem * const elem : input_mesh.active_element_ptr_range())
+    if (!block_ids.count(elem->subdomain_id()))
+      block_ids.insert(elem->subdomain_id());
+  // create depletion IDs
+  std::unordered_map<dof_id_type, dof_id_type> depl_ids =
+      MooseMeshUtils::getExtraIDUniqueCombinationMap(input_mesh, block_ids, id_names);
```
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/ReactorGeometryMeshBuilderBase.C:313]: document the convention for the ordering of the depltion ids

File: modules/reactor/src/meshgenerators/ReactorGeometryMeshBuilderBase.C [hunk 5]
```diff
@@ -254,3 +268,50 @@ ReactorGeometryMeshBuilderBase::print2dMetadataToConsole(const std::string metad
...
+    else
+      paramError("depletion_id_type",
+                 "'pin' or 'pin_type' is not allowed in depletion ID generation at assembly level");
+  }
+  else if (generation_level == DepletionIDGenerationLevel::Pin)
+    mooseError("Depletion ID generation is not supported at pin level yet in RGMB");
+  if (extrude)
+    id_names.push_back("plane_id");
+  id_names.push_back("region_id");
+  // prepare list of blocks
```
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/ReactorGeometryMeshBuilderBase.C:306]: do we not have a getter for the list of blocks

Comment (GiudGiud) [reply modules/reactor/src/meshgenerators/ReactorGeometryMeshBuilderBase.C:306]: also why are you block-restricting to all blocks? the routine call after will be faster if you pass an empty set (meaning not block-restricted)

File: modules/reactor/include/meshgenerators/ReactorGeometryMeshBuilderBase.h [hunk 6]
```diff
@@ -187,6 +189,25 @@ class ReactorGeometryMeshBuilderBase : public MeshGenerator
...
+    Assembly,
+    Core
+  };
+
+  /**
+   * add depletion IDs
+   * @param input_mesh input mesh
+   * @param option option for specifying level of details
+   * @param generation_level depletion id is generated at which reactor generator level
+   * @param extrude whether input mesh is extruded
```
Comment (GiudGiud) [inline modules/reactor/include/meshgenerators/ReactorGeometryMeshBuilderBase.h:205]: add a note that says what it does if the mesh is extruded

**Review by GiudGiud — CHANGES_REQUESTED** [review]

**Comment by yjung-anl** [comment]
@GiudGiud  Thank you! I will quickly address your comments.

**Comment by yjung-anl** [comment]
@GiudGiud 
Review comments have been addressed.

File: modules/reactor/src/meshgenerators/ReactorGeometryMeshBuilderBase.C [hunk 7]
```diff
@@ -254,3 +268,50 @@ ReactorGeometryMeshBuilderBase::print2dMetadataToConsole(const std::string metad
...
+  }
+  else if (generation_level == DepletionIDGenerationLevel::Assembly)
+  {
+    if (option == "pin")
+      id_names.push_back("pin_id");
+    else if (option == "pin_type")
+      id_names.push_back("pin_type_id");
+    else
+      paramError("depletion_id_type",
+                 "'pin' or 'pin_type' is not allowed in depletion ID generation at assembly level");
```
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/ReactorGeometryMeshBuilderBase.C:301]: do you mean assembly is not allowed?

Comment (yjung-anl) [reply modules/reactor/src/meshgenerators/ReactorGeometryMeshBuilderBase.C:301]: Oh it's mistake. Updated as: 
```c++
      paramError("depletion_id_type",
                 "'assembly_id' or 'assembly_type_id' is not allowd in depletion ID generation at "
                 "assembly level");
```

**Review by GiudGiud — APPROVED** [review]
Failures unrelated

**Comment by yjung-anl** [comment]
Thank you!!

**Comment by GiudGiud** [comment]
All good thanks for the contribution!

**Pull request merged**